_source/_posts/2021-08-13-reactive-java.md

```diff
@@ -0,0 +1,791 @@
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "2" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "4" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "2" [subscribe5]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "4" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "3" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "5" [map]
+2021-07-15 22:57:33.445 - INFO [parallel-4] - element "5" [subscribe4]
+2021-07-15 22:57:33.445 - INFO [parallel-1] - element "3" [subscribe5]
+2021-07-15 22:57:33.442 - INFO [parallel-3] - element "4" [subscribe3]
+2021-07-15 22:57:33.450 - INFO [parallel-3] - element "5" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-3] - element "5" [subscribe3]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "4" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "4" [subscribe5]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "5" [map]
+2021-07-15 22:57:33.450 - INFO [parallel-1] - element "5" [subscribe5]
+```
+
+As you can see, different subscription executions interleave, and as I am testing with only four CPU cores, four workers seem to be available: _parallel-1_, _parallel-2_, _parallel-3_, and _parallel-4_. The `subscription5` operations, execute in _parallel-1_.
+
+Again, observe how all operations for a given subscription are executed in the same thread.
+
+### Using Both `publishOn` and `subscribeOn`
+
+When the `publishOn` and `subscribeOn` operators are used, the `subscribe` call accepts a consumer. The consumer will execute in the context selected by `publishOn`, which defines the scheduler for downstream processing, including the consumer execution. At first this might be confusing to understand, so check the following example:
+
+```java
+@Test
+public void subscribeOnWithPublishOnTest() throws InterruptedException {
+    Flux.range(1, 3)
+            .map(v -> debug(v, "map1"))
+            .publishOn(Schedulers.parallel())
+            .map(v -> debug(v * 100, "map2"))
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe(w -> debug(w,"subscribe"));
+
+    Thread.sleep(5000);
+}
+```
+
+The log should look like the following lines:
+
+```
+2021-07-29 08:40:30.583 - INFO [boundedElastic-1] - element "1" [map1]
+2021-07-29 08:40:30.584 - INFO [parallel-1] - element "100" [map2]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "100" [subscribe]
+2021-07-29 08:40:30.585 - INFO [boundedElastic-1] - element "2" [map1]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "200" [map2]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "200" [subscribe]
+...
+```
+
+As you can see, the first `map` operator executes in the _boundedElastic-1_ thread, and the second `map` operator and the `subscribe` consumer execute in the _parallel-1_ thread. Notice that the `subscribeOn` operator is invoked after `publishOn`, but it affects the root of the chain and the operators preceding `publishOn` anyways.
```

Comment:
```suggestion
As you can see, the first `map` operator executes in the _boundedElastic-1_ thread, the second `map` operator and the `subscribe` consumer execute in the _parallel-1_ thread. Notice that the `subscribeOn` operator is invoked after `publishOn`, but it affects the root of the chain and the operators preceding `publishOn` anyways.
```
removed an 'and' to avoid a run-on...